pymmcore_plus/mda/_engine.py

```diff
@@ -0,0 +1,113 @@
+import time
+from typing import TYPE_CHECKING, Protocol, runtime_checkable
+
+from loguru import logger
+from useq import MDASequence
+
+from .events import PMDASignaler, _get_auto_MDA_callback_class
+
+if TYPE_CHECKING:
+    from ..core import CMMCorePlus
+
+
+@runtime_checkable
+class PMDAEngine(Protocol):
+    events: PMDASignaler
+
+    def cancel(self):
+        ...
+
+    def toggle_pause(self):
+        ...
+
+    def run(self, sequence: MDASequence):
+        ...
+
+
+class MDAEngine:
+    def __init__(self, mmc: "CMMCorePlus") -> None:
+        self.events = _get_auto_MDA_callback_class()()
+        # we can't use the Instance approach here
```

Comment:
if we really wanted to, we could set self._mmc to `None`  and then have a property that gets the global instance

```
@property
def mmc(self):
    return self._mmc or CMMCorePlus.instance()
```

Reply:
The other way I think is to call `instance` in the `run` method which is the only place it actually gets used.

Reply:
yep, that works too!

Reply:
or a third option of the `core.mda` being lazily created.

Reply:
I think I like keeping this in mda.  I like the optional input in the `__init__`... whether you use the `@property` for `mmc`, or just use `self._mmc or CMMCorePlus.instance()` in the run method, I don't mind